samples/net/openthread/coap_server/src/ot_coap_utils.c

```diff
@@ -0,0 +1,247 @@
+			server_context.on_provisioning_request();
+			server_context.provisioning_enabled = false;
+		}
+	}
+}
+
+static void light_request_handler(void *context, otMessage *message,
+				  const otMessageInfo *message_info)
+{
+	uint8_t command;
+
+	ARG_UNUSED(context);
+
+	if (otCoapMessageGetType(message) != OT_COAP_TYPE_NON_CONFIRMABLE) {
+		LOG_ERR("Light handler - Unexpected type of message");
+		goto end;
+	}
+
+	if (otCoapMessageGetCode(message) != OT_COAP_CODE_PUT) {
+		LOG_ERR("Light handler - Unexpected CoAP code");
+		goto end;
+	}
+
+	if (otMessageRead(message, otMessageGetOffset(message), &command, 1) !=
+	    1) {
+		LOG_ERR("Light handler - Missing light command");
+		goto end;
+	}
+
+	LOG_INF("Received light request: %c", command);
+
+	server_context.on_light_request((light_command)command);
+
+end:
+	return;
+}
+
+static void coap_default_handler(void *context, otMessage *message,
+				 const otMessageInfo *message_info)
+{
+	ARG_UNUSED(context);
+	ARG_UNUSED(message);
+	ARG_UNUSED(message_info);
+
+	LOG_INF("Received CoAP message that does not match any request or resource");
+}
+
+void ot_coap_activate_provisioning()
+{
+	server_context.provisioning_enabled = true;
+}
+
+void ot_coap_deactivate_provisioning()
+{
+	server_context.provisioning_enabled = false;
+}
+
+bool ot_coap_is_provisioning_active()
+{
+	return server_context.provisioning_enabled;
+}
+
+int ot_coap_init(otStateChangedCallback on_state_changed,
+		 provisioning_request_callback on_provisioning_request,
+		 light_request_callback on_light_request)
+{
+	otError error;
+
+	server_context.provisioning_enabled = false;
+	server_context.on_provisioning_request = on_provisioning_request;
+	server_context.on_light_request = on_light_request;
+
+	server_context.ot = get_ot_instance();
+	if (!server_context.ot) {
+		LOG_ERR("There is no valid OpenThread instance");
+		error = OT_ERROR_FAILED;
+		goto end;
+	}
+
+	provisioning_resource.mContext = server_context.ot;
+	provisioning_resource.mHandler = provisioning_request_handler;
+
+	light_resource.mContext = server_context.ot;
+	light_resource.mHandler = light_request_handler;
+
+	error = otSetStateChangedCallback(server_context.ot, on_state_changed,
+					  server_context.ot);
+	if (error != OT_ERROR_NONE) {
+		LOG_ERR("Failed to set OpenThread callback. Error: %d", error);
+		goto end;
+	}
+
+	otCoapSetDefaultHandler(server_context.ot, coap_default_handler, NULL);
+
+	error = otCoapAddResource(server_context.ot, &light_resource);
+	if (error != OT_ERROR_NONE) {
+		LOG_ERR("Failed to add CoAP resources. Error: %d", error);
+		goto end;
+	}
+
+	error = otCoapAddResource(server_context.ot, &provisioning_resource);
+	if (error != OT_ERROR_NONE) {
+		LOG_ERR("Failed to add CoAP resources. Error: %d", error);
+		goto end;
+	}
+
+	error = otCoapStart(server_context.ot, COAP_PORT);
+	if (error != OT_ERROR_NONE) {
+		LOG_ERR("Failed to start OT CoAP. Error: %d", error);
+		goto end;
+	}
+
+end:
+	return error;
```

Comment:
return type is int, but error is of type otError. Need to change return type or error type. Not sure if returning otError in non-openthread code is a good idea.